### PR TITLE
Add Bareflank Boxy / MicroV support

### PR DESCRIPTION
**Pull request by chp-io**
This PR adds support for [Bareflank boxy](https://github.com/Bareflank/boxy).
In order to be able to use this PR, I will also send a PR to the Bareflank Boxy repo for:
- Boxy VMM: Adding first class support for LibVMI in Boxy based on the [previous Bareflank extension](https://github.com/libvmi/bareflank_libvmi_extension) work.
- vmilinux: An example VM that includes a vmi init app and uses Buildroot to cross-compile LibVMI with this PR and build the rootfs.

This last point requires this PR to be accepted first.

### Bareflank Boxy

Bareflank is a very interesting hypervisor with multi-platform support (e.g. Windows, Linux and UEFI). It can boot from UEFI, or with its late-lunch feature, it can rootkit the current running host OS and demote it into a VM (dom0). With Boxy, an extension of the base hypervisor that adds guest support, we can run LibVMI into a tiny linux VM and, for the first time, have libvmi introspect a **Windows host**!
Of course, we can also introspect a Linux host just like before. Another difference is that the libvmi process isn't directly scheduled by the host OS and it won't show up in the process list.
Instead we see bfexec which is Boxy's way to use the host scheduler to schedule VMs. i.e. When bfexec is scheduled by the host OS, it then vmcalls the VMM to execute its associated VM.

#### LibVMI with the Bareflank base hypervisor only

When running without Boxy (i.e. with the base hypervisor only) LibVMI runs directly on the host it is introspecting which limits it to Linux. As shown in the [original bareflank driver PR](https://github.com/libvmi/libvmi/pull/693), notice the libvmi process `vmi-process-list` actually displayed as part of the process list running on the host:

```
$ ~/workspace/libvmi$ ./examples/vmi-process-list host
Process listing for VM (null) (id=0)
[    0] swapper/0 (struct addr:ffffffff9e212480)
...
[ 4841] vmi-process-lis (struct addr:ffff9293396f9740)
```

#### LibVMI with Bareflank Boxy

With Boxy, libvmi runs inside a VM with an ultra-light Linux. Bfexec can pass arguments to the linux cmdline so I made an init app that parses the kernel command line and passes it to the VMI application (e.g. `vmi-process-list`). In this example, we're introspecting a Windows host from the host in cygwin. With bfexec's UART emulation, we can see linux boot, start init which in turn starts the vmi process list example, all done in milliseconds:

```
$ ./bareflank/prefixes/x86_64-userspace-pe/bin/bfexec.exe --bzimage --path ./buildroot/output/images/bzImage --initrd ./buildroot/output/images/rootfs.cpio.gz --uart=0x3f8 --verbose --size=0x8000000 --cmdline="vmi=/usr/bin/vmi-process-list,-n,dom0"

Created VM from bzImage file:
--------------------------------------------------------------------------------
    kernel | ./buildroot/output/images/bzImage
    initrd | ./buildroot/output/images/rootfs.cpio.gz
 domain id | 1
  ram size | 128MB
   cmdline | console=uart,io,0x3f8,115200n8 vmi=/usr/bin/vmi-process-list,-n,dom0

[0xOutput from VM's UART:
10000--------------------------------------------------------------------------------
]
vcpu fault: [    0.000000] Linux version 4.20.0-rc7 (vagrant@vagrant) (gcc version 8.4.0 (Buildroot 2020.02.2)) #1 Wed Jun 10 05:54:59 UTC 2020
[    0.000000] Command line: console=uart,io,0x3f8,115200n8 vmi=/usr/bin/vmi-process-list,-n,dom0
[    0.000000] KERNEL supported cpus:
[    0.000000]   Intel GenuineIntel
[    0.000000] x86/fpu: x87 FPU will use FXSAVE
[    0.000000] BIOS-provided physical RAM map:
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x00000000000e7fff] usable
[    0.000000] BIOS-e820: [mem 0x00000000000e8000-0x00000000000fffff] reserved
[    0.000000] BIOS-e820: [mem 0x0000000000100000-0x00000000080fffff] usable
[    0.000000] BIOS-e820: [mem 0x00000000fec00000-0x00000000fffffffe] reserved
[    0.000000] earlycon: uart0 at I/O port 0x3f8 (options '115200n8')
[    0.000000] printk: bootconsole [uart0] enabled
[    0.000000] NX (Execute Disable) protection: active
[    0.000000] Hypervisor detected: Boxy Hypervisor
[    0.000001] clocksource: boxy-clocksource: mask: 0xffffffffffffffff max_cycles: 0x255cb6cc5db, max_idle_ns: 440795203504 ns
[    0.000390] tsc: Detected 2592.000 MHz processor
...
[    0.088346] Run /sbin/init as init process
[    0.088700] [U] Init: reached
[    0.088851] [U] Init: Starting /usr/bin/vmi-process-list -n dom0 at Sat Jun 13 23:36:09 2020
[    0.100063] random: vmi-process-lis: uninitialized urandom read (4 bytes read)
[    0.147687] [U] Process listing for VM (null) (id=0)
[    0.148272] [U] [    4] System (struct addr:ffff8a09392c1140)
[    0.148784] [U] [  148] Registry (struct addr:ffff8a09393f3040)
[    0.149300] [U] [  516] smss.exe (struct addr:ffff8a0941deb040)
[    0.149816] [U] [  840] csrss.exe (struct addr:ffff8a0945135140)
[    0.150337] [U] [  968] wininit.exe (struct addr:ffff8a0947c240c0)
[    0.150862] [U] [  972] csrss.exe (struct addr:ffff8a0947c27140)
[    0.151379] [U] [  468] services.exe (struct addr:ffff8a0947c60280)
[    0.151920] [U] [  636] WerFault.exe (struct addr:ffff8a0947d042c0)
[    0.152449] [U] [  624] lsass.exe (struct addr:ffff8a0947d06280)
[    0.152972] [U] [ 1052] svchost.exe (struct addr:ffff8a0947d95080)
[    0.153592] [U] [ 1076] svchost.exe (struct addr:ffff8a0947d980c0)
[    0.154120] [U] [ 1100] fontdrvhost.ex (struct addr:ffff8a0947d9f340)
[    0.154654] [U] [ 1168] svchost.exe (struct addr:ffff8a0948467240)
[    0.155179] [U] [ 1216] svchost.exe (struct addr:ffff8a094848d080)
[    0.155709] [U] [ 1280] winlogon.exe (struct addr:ffff8a09485e20c0)
[    0.156240] [U] [ 1332] WUDFHost.exe (struct addr:ffff8a0948613240)
[    0.156773] [U] [ 1372] fontdrvhost.ex (struct addr:ffff8a09486d3340)
[    0.157312] [U] [ 1444] WUDFHost.exe (struct addr:ffff8a0948716080)
[    0.157842] [U] [ 1496] LogonUI.exe (struct addr:ffff8a0948720280)
[    0.158369] [U] [ 1504] dwm.exe (struct addr:ffff8a0948722080)
...
[    0.285676] [U] [11872] firefox.exe (struct addr:ffff8a094daec080)
[    0.286203] [U] [16148] audiodg.exe (struct addr:ffff8a094a5df080)
[    0.286775] [U] [ 8316] mspdbsrv.exe (struct addr:ffff8a095031a080)
[    0.287306] [U] [13420] svchost.exe (struct addr:ffff8a095401c080)
[    0.287829] [U] [ 6640] svchost.exe (struct addr:ffff8a094fca6080)
[    0.288353] [U] [ 8012] svchost.exe (struct addr:ffff8a095425f240)
[    0.288900] [U] [ 6320] cpptools-srv.e (struct addr:ffff8a09534ef080)
[    0.289438] [U] [  740] SearchProtocol (struct addr:ffff8a094df27080)
[    0.289972] [U] [12768] SearchFilterHo (struct addr:ffff8a094a50b340)
[    0.290509] [U] [ 4672] zsh.exe (struct addr:ffff8a095008d080)
[    0.291046] [U] [ 8020] bfexec.exe (struct addr:ffff8a0952feb080)
[    0.308360] [U] Init: return code is 0. Nothing left to do.
[    0.308590] reboot: System halted
```

### CODE REVIEW

**Comment by chp-io** [comment]
CC: @tklengyel @rianquinn

**Comment by tklengyel** [comment]
@chp-io I looked through, mostly looks fine, some questions I have about the kpcr_find2 adjustments but first please address all CI issues

File: libvmi/driver/bareflank/bareflank.c [hunk 1]
```diff
@@ -68,7 +68,7 @@ static inline uint64_t parse_reg_value(const char *reg, json_object *root)
     // parse the json and get the value of the key
     json_object *return_obj = NULL;
     json_object_object_get_ex(root,reg,&return_obj);
-    return json_object_get_int64(return_obj);
+    return (uint64_t) json_object_get_uint64(return_obj);
```
Comment (chp-io) [inline libvmi/driver/bareflank/bareflank.c:71]: This is a new API introduced in **json-c 0.14** allowing us to not have to split `uin64_t` values into high/low values anymore.

**Comment by chp-io** [comment]
Travis doesn't like that I am using a newer version of **json-c**. Travis would need to install [json-c 0.14](https://github.com/json-c/json-c/releases).

**Comment by chp-io** [comment]
Also note that LibVMI with Boxy is built within a Buildroot environment which I have control over. The host doesn't need to have json-c installed, Buildroot takes care of this.
Should modify Travis to install `json-c 0.14` from the gziped version or should I simply disable Bareflank for now until 0.14 becomes available from the package manager?

File: libvmi/driver/bareflank/bareflank.c (hunk 1 above)
Comment (tklengyel) [reply libvmi/driver/bareflank/bareflank.c:71]: Well, if you rely on a particular API being available you better check for that during the configure stage to alert the user that they need to install a newer version of the lib. Especially if its not something available by default in the repository.

**Comment by tklengyel** [comment]
>  Should modify Travis to install json-c 0.14 from the gziped version or should I simply disable Bareflank for now until 0.14 becomes available from the package manager?

No, we actually had this issue with json-c for a long time so I would prefer adding a json-c submodule pointing to https://github.com/json-c/json-c/releases/tag/json-c-0.14-20200419 to LibVMI and updating the install instructions to build and install it from there instead of using the package manager. Once the LTS Ubuntu and debian stable gets the package updated we can remove the submodule but till then we'll make due.

**Comment by tklengyel** [comment]
Only caveat to that is the build needs to continue to work with old version of json-c as well, so just have to check whether the new version is available and only build Bareflank/Boxy if it is.

**Comment by tklengyel** [comment]
Also, once the submodule is added you can update travis to build and install json-c from there so we can at least compile-test the Bareflank/Boxy driver.

**Comment by chp-io** [comment]
Got it!

**Comment by tklengyel** [comment]
@chp-io Please also update `configure.ac` to check the json-c version

**Comment by chp-io** [comment]
I forced pushed to squash the last commit onto the previous one and keep the history clean.

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

**Comment by chp-io** [comment]
Commit https://github.com/libvmi/libvmi/pull/909/commits/cebf2b728cd16576166768e4cf5db1210eab323e adds support for the new [MicroV specification](https://github.com/Bareflank/MicroV/blob/88e9b429e58c43fe7dad9bdf589716a81c5ddb02/docs/MicroV%20VM%20Specification.md).
It also adds better support for GP and MSR registers.

The new MicroV implementation doesn't use JSON anymore so I reverted the json-c 0.14 build checks.

Since json-c 0.14 isn't used anywhere anymore, should I remove it completely and we can re-add it once we actually need it again? It currently still exists in the git submodule and travis.

**Pull request closed without merging**